src/edit.rs

```diff
@@ -217,6 +218,46 @@ impl<'out, 'prompt, H: Helper> State<'out, 'prompt, H> {
     pub fn is_default_prompt(&self) -> bool {
         self.layout.default_prompt
     }
+
+    pub fn validate(&mut self) -> Result<bool> {
+        if let Some(validator) = self.helper {
+            self.changes.borrow_mut().begin();
+            let result = validator.validate(&mut ValidationContext::new(self))?;
+            let corrected = self.changes.borrow_mut().end();
+            let validated = match result {
+                ValidationResult::Incomplete => {
+                    self.edit_move_end()?;
+                    self.edit_insert('\n', 1)?;
+                    false
+                }
+                ValidationResult::Valid(msg) => {
+                    // Accept the line regardless of where the cursor is.
+                    self.edit_move_end()?;
+                    if corrected || self.has_hint() || msg.is_some() {
+                        // Force a refresh without hints to leave the previous
+                        // line as the user typed it after a newline.
+                        self.refresh_line_with_msg(msg)?;
+                    }
+                    true
+                }
+                ValidationResult::Invalid(msg) => {
+                    if corrected || self.has_hint() || msg.is_some() {
+                        self.refresh_line_with_msg(msg)?;
```

Comment:
So the message here is highlighted the same as a hint, right? This should be documented in the Highlighter trait.